Change esp32.RMT.source_freq() to class method

bulletmark opened the pull request:
To create a `esp32.RMT()` instance with an optimum (i.e. highest resolution) `clock_div` is currently awkward because you need to know the source clock frequency to calculate the best `clock_div` but unfortunately that is only currently available as an `source_freq()` method on the instance **after** you have already created it, i.e. we have a chicken and egg problem. So `RMT.source_freq()` should really be a class method, not an instance method and this PR changes that in the code and documentation. Of course this change is backwards compatible for existing code because you can still reference that function from an instance, or now also, from the class, i.e.

```python
from esp32 import RMT
rmt = RMT(0, pin)

assert rmt.source_freq() == RMT.source_freq()
````

bulletmark committed:
esp32/esp32_rmt: Change RMT.source_freq() to class method.

To create an esp32.RMT() instance with an optimum (i.e. highest resolution)
clock_div is currently awkward because you need to know the source clock
frequency to calculate the best clock_div, but unfortunately that is only
currently available as an source_freq() method on the instance after you
have already created it.  So RMT.source_freq() should really be a class
method, not an instance method.  This change is backwards compatible for
existing code because you can still reference that function from an
instance, or now also, from the class.

Signed-off-by: Mark Blakeney <mark.blakeney@bullet-systems.net>

mattytrentini commented:
That looks to be a neat solution, looks good to me. 👍🏻

bulletmark commented:
@mattytrentini I know you are the original author of this and since you are here - to compute the optimum `clock_div` you also need the duration value which is 32767 units. You state this number in the documentation but it is not available via the API so should I add it with this PR? Do you prefer it as a class constant or another class method?

IhorNehrutsa commented:
@bulletmark 
I think you don't need to add 32767 constant to class due the memory usage minimization. 
For example, UART and SPI baud rates are not class constants, but simple integer arguments.

bulletmark commented:
When you are setting up a `RMT.write_pulses(duration, data=True)` call then you have to make sure that each element in your `duration` list is `<= 32767` since [the Espressif C API allows only a maximum 15 bit unsigned int](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/rmt.html#layout-of-rmt-symbols) for the `period` value. The [MP documentation](https://docs.micropython.org/en/latest/library/esp32.html#esp32.RMT.write_pulses) does state this. If you have a variable frequency and pulse width, and you want to optimize pulse resolution, then you must do a calculation beforehand to ensure you normalize the array to keep all list values lower than that. That calculation requires `RMT.source_freq()`, `RMT.clock_div()`, and this magic `32767` value. Just seems to me that magic value should be `RMT.pulses_duration_max` and not have to be defined as an explicit `const` in my code.

dpgeorge commented:
> Just seems to me that magic value should be `RMT.pulses_duration_max` and not have to be defined as an explicit `const` in my code.

Yes I agree, it would make sense to add this constant.

> For example, UART and SPI baud rates are not class constants, but simple integer arguments.

No, this is different, it's not a set of constants for valid rates, it's just a single constant for the limit of the arguments provided to `RMT.write_pulses()`

So, @bulletmark , feel free to add such a constant, as a second commit in this PR, if you like.  I suggest calling it `RMT.PULSE_MAX`.

bulletmark committed:
esp32/esp32_rmt: Add RMT.PULSE_MAX constant.

If you have a variable frequency and pulse width, and you want to optimize
pulse resolution, then you must do a calculation beforehand to ensure you
normalize the array to keep all list values within bound.  That calculation
requires RMT.source_freq(), RMT.clock_div(), and this 32767 constant.

Signed-off-by: Mark Blakeney <mark.blakeney@bullet-systems.net>

dpgeorge commented:
Thanks for updating.  Now merged.
